### ISSUE TITLE
[Discussion] Mitigations to improve peers privacy

### ISSUE THREAD

**Issue by Chocobozzz** [webtorrent/bittorrent-tracker#271]
Hi :)

I would like to discuss the possibility (or not) to improve peers privacy (related to the IP) when using the `bittorrent-tracker` package, because there are some concerns about viewers privacy in [PeerTube](https://github.com/Chocobozzz/PeerTube). I came to different possible mitigations, but since I'm not a BitTorrent expert I would like to have some feedbacks:

 * Have a daemon (or something like that) announcing some torrents with fake IPs. People that would want to spy a specific torrent (with an `announce`), could not differentiate real peers than fake ones. They would need to connect directly to the peer to check if it is real one or not -> more work for the spy. And even with that, if they cannot connect to a peer, they would not be sure at 100% this is a fake one because some real peers cannot use P2P with WebRTC (network limitations).
 * Modify the max announce peers limit to be very small (~10/20)
 * Limit the number of `announce` requests by a specific IP per minute (for example limit to 5 `announce` per minute?)

Any thoughts on this?

Thanks <3

**Comment by DiegoRBaquero** [comment]
1. I don't like the first or see it as a reasonable solution.
2. Could be an option
3. Should be an option

my 2c

**Comment by Chocobozzz** [comment]
Thanks @DiegoRBaquero for the feedback. Could your elaborate why you don't like the first point?

**Comment by rigelk** [comment]
@Chocobozzz if I may elaborate on 1., this has been in place for quite some time at some of the major trackers and dates back to [as early as 2008](https://torrentfreak.com/the-pirate-bay-tricks-anti-pirates-with-fake-peers-081020/):

> Polluting the evidence works like this. When a client asks for a list of peers who are downloading the same torrent, the tracker software automatically inserts several “random IP addresses” that are not in the swarm. They are based on existing sub-nets, but might be from people who may not even be aware that BitTorrent exists. This means that the evidence that’s being gathered by anti-piracy companies includes IPs that belong to people that were not downloading the movie or album they are accused of. 

But that only works so long as there is no actual proof of the transfer. Protecting from the latter is more tricky, and − as far as I know − requires blocklists, which are already out of scope of this issue. A paper from 2012, [The Unbearable Lightness of Monitoring: Direct
Monitoring in BitTorrent](http://www.cs.bham.ac.uk/~tpc/Papers/P2PMonitor.pdf), shows their efficiency leaves much to be desired (~68% efficiency).

Anything beyond requires using a different protocol than BitTorrent. Apparently that's what [some guy did in an obscure research](http://archive.is/ZQLkL) in 2015, showing only a low percentage of nodes are actually initiating a content transfer so as to prove the client is authentic. Interesting but out of scope.

**Comment by davalapar** [comment]
Hi @Chocobozzz , privacy is a pretty broad topic but if you feel like experimenting, you may try out concepts from [this](https://getmonero.org/resources/research-lab/) and [this](https://latacora.singles/2018/04/03/cryptographic-right-answers.html). I don't know the ins and outs of PeerTube per se, but I can say there's only so much you can do to seriously address this without drastically affecting your end-users' experience.

**Comment by iamhansen** [comment]
Option 1 has been implemented in opentracker. I think it should be added because no one is being stopped to just put in a script that announces the visitor's ip to an info hash that an antipiracy corporation is monitoring.

**Comment by Chocobozzz** [comment]
@kelu1018 I'm not an expert in the webtorrent protocol, but I think you are able to find the IP address in the `SDP offer` (between `tracker` and `peer 2`). Tracking users using webtorrent protocol is harder than the classic bittorrent protocol, because the tracker does not send directly IP address to those who request them. But it's still possible if you seed a particular file: you just have to wait the `SDP offer` from the tracker. 

> For the users of our website it is of course very good if the IP address is not so easy to find out while watching a video on PeerTube

Users can already disable P2P in their settings, and I'll try to implement a *disable P2P* option in the future for instance administrators who don't have a lot of visitors and prefer to improve their users privacy.

**Comment by DontUseGithub** [comment]
@Chocobozzz Can you please update your first post, if I'm not wrong, 1. has already been implemented. Thank you.

**Comment by Chocobozzz** [comment]
This is just a discussion, not a checklist of things implemented in PeerTube.

**Comment by toogley** [comment]
The first step to improve the situation would be to make that configurable in the user-account. That means, an user should be able to decide whether they want to use a HTTP or a peer 2 peer connection. This setting could be saved in the user settings, which means a user has to be logged in.

This is an improvement, because it improves the user experience for Tor users for instance (for which peer2peer connections are a problem, i think).

The next step would be to make that configurable for non-loggedin users. Currently, the situation is:
* peer2peer is enabled by default after installation
* automatically start playing/loading the video is on by default after installation
* a web-banner was previously shown to indicate that ihe IP-Address is revealed. This is 

i think it would be sensible to change that to:
* HTTP connection is used by default after installation (making peer2peer technolgy opt-in for instance admins). This is better because unaware users might not think of their IP-Address to be revealed when they install their youtube-alternative. Maybe they haven't read about peer2peer before. Maybe they didn't thought about the consequence of their IP-Address being revealed.
* the video is still played automatically, which is not harming anymore because HTTP does not reveal the IP-address.
* a FAQ entry on "what to do with high load" is added which explains that you can use peer2peer technology to reduce server bandwith.
* regardless of being logged in, users can enable peer2peer connections on a video. This setting saved, so further connections to a different video are also made using peer2peer technolgy. Which means, users can enable peer2peer technology both in the account settings and for one concrete video.

An especially useful and user friendly thing to do would be:
* when peer2peer is enabled somewhere and a user starts watching a video, the connection is started using HTTP. The user is asked with an not-annoying message whether they want to use peer2peer technology, with a message that this reveals their IP-address. This also prevents an accidential IP-leakage, since the user has to conciously do something. This could be saved in a cookie or user-setting for further videos, so the user has to do that only once.


 think this is the most sensible thing to do.

**Comment by toogley** [comment]
> Have a daemon (or something like that) announcing some torrents with fake IPs. People that would want to spy a specific torrent (with an announce), could not differentiate real peers than fake ones. They would need to connect directly to the peer to check if it is real one or not -> more work for the spy. And even with that, if they cannot connect to a peer, they would not be sure at 100% this is a fake one because some real peers cannot use P2P with WebRTC (network limitations).

This is not solving one important use-case. Lets say, you are a investigative journalist working in turkey, russia, the USA, whatever. You want to report on local occurrences, but you don't want to be known by the police or so. Maybe you use tor/tails, maybe not. Maybe the stuff you want to research about does use cloudflare, so you can't use tor properly.

Nevertheless, you have found a peertube instance of local protestants. they have uploaded videos in which they criticize the government on something. You want to watch this video to report on this topic.

Which means, this person is:
* not a tech person. Maybe they know what an ip-address is, because they read about how to protect themselves. But they don't have tech experience, which means they have not the knowledge to look out for potential things which harm their privacy. I am a tech-person. I know, when i read "peer to peer" or something like that, this will reveal my ip-address. This investigative jorunalist also knows that probably, but only after carefully thinking about it. So it is very easy for this person to accidentially reveal the IP-address because they carelessly clicked onto a video. 
* this investigative journalist probably doesn't know how to prevent this IP-leakge, *unless* a clearly understandable asks this person if they want to leak their IP-address.
* even I as a tech person have already accidentially leaked my IP-address and only noticed it slightly afterwards.

And yes, you could say, that this is the responsibility of the instance-owner or the account owner. They would have to disable peer2peer and only use HTTP because they know they will be a target for the police to investigate. They know the police is interested in learning the IP-addresses of every person who watches this video.

I agree, but not fully. i think it is very important that users have the freedom to decide for themselves what level of privacy they need. Investigative journalists will need it.

**Comment by toogley** [comment]
Additional motivation to fix this issue:

*As you already noticed, Extinction Rebellion has  [also a peertube instance](https://tube.rebellion.global). I want to point out, that UK considers to declare [extinction rebellion a criminal organization](https://www.euronews.com/living/2020/09/07/extinction-rebellion-responds-to-ridiculous-criminalisation-threat), which could lead up to 5 years in jail. And i guess its no surprise that a big part of the persons viewing the Extinction rebellion videos are the activists themselves. Which means, since the IP-Addresses of the peer2peer feature are "public information", the british police can easily obtain these private information, even without a court order. Which means, i think the activists would be generally interested to protect their information the same way as the investiagtive journalist above wants.
* Additionally, i want to point out that extinction rebellion is also active in areas where human rights are endangered, for instance [qatar](https://rebellion.global/groups/qa-qatar/), [china](https://rebellion.global/groups/cn-china/) or [russia](https://rebellion.global/groups/ru-russia/). And the above peertube instance is a global instance, which means every branch can request an account.

Additionally, peertube is already used in anarchist groups, for instance [kolektiva.media](https://kolektiva.media/videos/local). see also [this description of the groups involved](https://channelzeronetwork.com/kolektiva-media-a-new-anarchist-videohosting-alternative/). And remember that donald trump, the president of the USA, has recently [ranted about the anarchists](https://www.huffpost.com/entry/trump-deadly-cans-of-soup_n_5f4fbcc6c5b69eb5c0379f01), which means people watching the videos are potential targets. And remember that the USA currently, doesn't care much about human rights and [uses Unmarked Vehicles To Grab People In Portland](https://text.npr.org/s.php?sId=892277592), without telling them the reason or how long they will be held. They don't have the possiblity to call lawyers.

Which means that for these users a proper fix for this feature would be very interesting, i guess.

Also, i want to point out that the people in the hambacher forst [also would be interested to use peertube and promote it, but don't precisely because of the lack of this feature](https://todon.nl/@hambibleibt/103550659625788483).

So in essence, with this writeup, i want to stand in solidarity to @hambibleibt (<3) and explain the background of  this privacy need.

**Comment by toogley** [comment]
Also notice, that using Tor or VPN is **Not** a secure alternative, since [webrtc](https://blog.ipvanish.com/webrtc-security-hole-leaks-real-ip-addresses/) can cause an IP-address leakage, which is very hard to determine if it had happened or not.

That requires the ability for users to always use HTTP-connections when they want privacy. peer2peer must not used under no circumstances.

**Comment by toogley** [comment]
ah, and i only now noticed that this is the wrong repository.

**Comment by anon97945** [comment]
Can't you switch off the "Help share videos being played" option by default and only make it available if the user wants to support the server. After all, the server owner would then decide for himself. Or is this option already available for server operators?
